### DISCUSSION TITLE
Custom Dataset

### DISCUSSION THREAD

**Discussion by Momods** [ultralytics/ultralytics#9869] [Q&A] [answered]
I am currently working on a project that involves using YOLO for object detection, where each class may have associated metadata that the trained model should output. I have access to a dataset consisting of images, corresponding labels, and metadata for each image. My objective is to modify the training process to incorporate both the images and their associated metadata as inputs to the model, enabling it to output not only the detected classes but also the associated precision and relevant metadata.

**Specific Questions:**

1. How can I adjust the YOLO training process to include both images and metadata as inputs to the model?
2. What modifications are needed in the model architecture and loss function to accommodate the inclusion of metadata during training?
3. Are there any specific data preprocessing steps required to ensure compatibility between image data and metadata inputs?
4. How can I access and utilize the output metadata provided by the trained YOLO model alongside the class predictions?

**Expected Outcome:**

By addressing these questions and adjusting the training process accordingly, I aim to develop a YOLO-based object detection model that not only accurately detects different classes but also delivers relevant metadata associated with each detection.

**Comment by glenn-jocher** [comment, marked as answer]
@Momods hello! Great to see your initiative to enhance object detection with metadata inclusion using YOLO. Here are some recommendations for your project's objectives:

1. **Including Images and Metadata as Inputs:** To feed both images and metadata into the model, you might consider a two-branch network architecture where one branch processes images (traditional YOLO) and the other branch processes the metadata. These branches can merge at a certain point in the architecture, combining features before detection layers. 

    ```python
    # Pseudo-code for a two-branch model architecture idea
    class MyYOLOWithMetadata(nn.Module):
        def __init__(self):
            super().__init__()
            self.image_branch = YOLOBranch()
            self.metadata_branch = MetadataBranch()
            # merge branches
            self.combined_features = CombineBranches()
        
        def forward(self, images, metadata):
            img_features = self.image_branch(images)
            meta_features = self.metadata_branch(metadata)
            combined = self.combined_features(img_features, meta_features)
            # Further processing
            return combined
    ```
   
2. **Model Architecture and Loss Function Adjustments:** The model architecture might need additional layers in both the image and metadata branches to process and merge the different types of data effectively. About the loss function, besides the traditional object detection losses, you might need to include a loss component that accounts for the metadata accuracy, hence a multi-task learning approach.

3. **Data Preprocessing:** For metadata, ensure it is normalized or encoded appropriately to maintain scale compatibility with neural network inputs. If metadata is categorical, consider one-hot encoding. For numerical metadata, normalization (e.g., to have zero mean and unit variance) could be helpful.

4. **Accessing and Utilizing Output Metadata:** The model's forward method should be adjusted to return not only class predictions but also the relevant metadata predictions. During inference, you can access these outputs directly from the model's prediction:

    ```python
    # Pseudo-code for using the model
    model = MyYOLOWithMetadata()
    image, metadata = prepare_data()
    class_preds, metadata_preds = model(image, metadata)
    ```
   
The expected outcome is feasible but requires careful design and implementation. I recommend focusing on how to effectively combine the image features with metadata for a unified representation that the model can use for effective learning and inference. Best of luck with your project! 🚀
